rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1451,6 +1451,67 @@ impl SystemState {
             .time_out_messages(current_time, own_canister_id, local_canisters)
     }
 
+    /// Enqueues "deadline expired" references for all expired best-effort callbacks
+    /// without a response.
+    ///
+    /// Returns `Err` if a `SystemState` internal inconsistency prevented  one or
+    /// more "deadline expired" references from being enqueued.
+    pub fn time_out_callbacks(
+        &mut self,
+        current_time: CoarseTime,
+        own_canister_id: &CanisterId,
+        local_canisters: &BTreeMap<CanisterId, CanisterState>,
+    ) -> Result<(), Vec<StateError>> {
+        if self.status == CanisterStatus::Stopped {
+            // Stopped canisters have no call context manager, so no callbacks.
+            return Ok(());
+        }
+
+        let aborted_or_paused_callback_id = self
+            .aborted_or_paused_response()
+            .map(|response| response.originator_reply_callback);
+
+        // Safe to unwrap because we just checked that the status is not `Stopped`.
+        let call_context_manager = call_context_manager_mut(&mut self.status).unwrap();
+
+        let mut errors = Vec::new();
+        let expired_callbacks = call_context_manager
+            .expire_callbacks(current_time)
+            .collect::<Vec<_>>();
+        for callback_id in expired_callbacks {
+            if Some(callback_id) == aborted_or_paused_callback_id {
```

Comment:
I now realize I hadn't thought about DTS earlier. Are we sure we'll never push a late response into an input queue if we previously pushed a timeout response, and that timeout response happens to trigger a DTS execution that's progressing exactly as the "real" response comes in?

Also, if a DTS execution gets aborted (e.g., because of a checkpoint), do we need to reinsert it into the `unexpired_callbacks` again?

Reply:
Hmm... You may have uncovered one very nasty bug.

We don't explicitly check against a possible aborted or paused response when inducting responses. I guess I was assuming that the respective `CallbackId` will stay in `CanisterQueues::callbacks_with_enqueued_response` until the callback is unregistered. But that's not the case: we remove it from that set as soon as the response is popped.

Regardless, even if we did explicitly check against na aborted or paused response execution, there would still be a loophole. While that would prevent inducting a response "duplicating" the already paused or aborted one, it would not prevent us from producing a reject response by timing out an outbound request. While I don't think that's an actual issue (because we first time out messages, then callbacks; so we would first produce an actual reject response, then try to produce a `DeadlibeExpired` compact response here), that's still not something I'm comfortable relying on.

But the somewhat hacky, yet safer way around that would be to remember the last `CallbackId` we popped; and include that alongside `CanisterQueues::callbacks_with_enqueued_response` in the duplicate response check.

I'll think about this some more, since it's such a nasty one. And thank you for asking the question.